internal/cache/file/cache_handler.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/data"
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/file/downloader"
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/lru"
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/util"
+	"github.com/googlecloudplatform/gcsfuse/internal/locker"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
+)
+
+// CacheHandler Responsible for managing fileInfoCache as well as fileDownloadManager.
+type CacheHandler struct {
+	// fileInfoCache contains the reference of fileInfo cache.
+	fileInfoCache *lru.Cache
+
+	// jobManager contains reference to a singleton jobManager.
+	jobManager *downloader.JobManager
+
+	// Directory location which would contain local file to keep the downloaded data.
+	cacheLocation string
+
+	// Guards the handling of cache entry insertion while creating the CacheHandle.
+	// Also guards post eviction logic E.g. cancelling and deletion of async download job,
+	// deletion of local file containing downloaded data.
+	mu locker.Locker
+}
+
+func NewCacheHandler(fileInfoCache *lru.Cache, jobManager *downloader.JobManager, cacheLocation string) *CacheHandler {
+	return &CacheHandler{
+		fileInfoCache: fileInfoCache,
+		jobManager:    jobManager,
+		cacheLocation: cacheLocation,
+		mu:            locker.New("FileCacheHandler", func() {}),
+	}
+}
+
+func (chr *CacheHandler) createLocalFileReadHandle(objectName string, bucketName string) (*os.File, error) {
+	fileSpec := data.FileSpec{
+		Path: util.GetDownloadPath(chr.cacheLocation, util.GetObjectPath(bucketName, objectName)),
+		Perm: util.DefaultFilePerm,
+	}
+
+	return util.CreateFile(fileSpec, os.O_RDONLY)
+}
+
+func (chr *CacheHandler) performPostEvictionWork(fileInfo *data.FileInfo) error {
+	key := fileInfo.Key
+	_, err := key.Key()
+	if err != nil {
+		return fmt.Errorf("error while performing post eviction: %v", err)
+	}
+
+	chr.jobManager.RemoveJob(key.ObjectName, key.BucketName)
```

Comment:
can we add comment here that Removing Job doesn't delete the job object itself but invalidates the job and hence it's possible that some existing cache handle will have reference to older job object but that job object will have INVALID status.